.github/api_guidelines.md

```diff
@@ -8,10 +8,132 @@ in slack channel [#forum-acs-api-design](https://redhat-internal.slack.com/archi
+  - [/proto/storage](https://github.com/stackrox/stackrox/blob/master/proto/storage) 
+  - [/proto/internalapi](https://github.com/stackrox/stackrox/blob/master/proto/internalapi) 
+  - [/proto/api/v1](https://github.com/stackrox/stackrox/blob/master/proto/api/v1)
+- All new services and methods defined in the package [/proto/api/v1](https://github.com/stackrox/stackrox/blob/master/proto/api/v1) 
+**must not** import from the following packages:
+  - [/proto/storage](https://github.com/stackrox/stackrox/blob/master/proto/storage)
+  - [/proto/internalapi](https://github.com/stackrox/stackrox/blob/master/proto/internalapi)
+  - [/proto/api/v2](https://github.com/stackrox/stackrox/blob/master/proto/api/v2)
+- All new messages defined in the package [/proto/internalapi](https://github.com/stackrox/stackrox/blob/master/proto/internalapi)
+  **must not** import from the following packages:
+  - [/proto/storage](https://github.com/stackrox/stackrox/blob/master/proto/storage)
+  - [/proto/api/v1](https://github.com/stackrox/stackrox/blob/master/proto/api/v1)
+  - [/proto/api/v2](https://github.com/stackrox/stackrox/blob/master/proto/api/v2)
+- All services of the type `APIServiceWithCustomRoutes` **must not** import from the following packages:
+  - [/proto/storage](https://github.com/stackrox/stackrox/blob/master/proto/storage) 
+  - [/proto/internalapi](https://github.com/stackrox/stackrox/blob/master/proto/internalapi)
+- Structs used by APIs, such as those defined in [/proto/api/](https://github.com/stackrox/stackrox/blob/master/proto/api/), 
+[/proto/internalapi](https://github.com/stackrox/stackrox/blob/master/proto/internalapi), **must not** be 
+written to database.
+
+Exposing internal data structures, especially the ones representing and affecting database schema directly at 
+the API can pose security risks. By carefully selecting and limiting the data exposed via the API, we can prevent 
+unauthorized access and potential data breaches which can have compliance ramifications. 
+
+Furthermore, exposing unnecessary data in the API can lead to overfetching, where the client receives more data 
+than required. This inefficiency increases network traffic, consumes extra bandwidth, and impacts API performance. 
+By limiting API data to what is actually needed by clients, we optimize resource usage and improve response times. 
+
+Consider the following database/internal structures as an example data structure used in Central.
+
+```
+Deployment {
+  string id
+  string name
+  repeated KeyValue labels
+  repeated KeyValue annotations
+  repeated Container containers
+}
+  
+Container {
+  string name
+  string image_name
+  repeated Volume volumes
+}
+```
+
+Exposing the above data structure in the API will always lead to reading the all `Container` bytes from the database, 
+even if the user does not need them. Instead, design the API such that users can request the information 
+they need, as below:
```

Comment:
This might be nitpicking, but I don't really agree with the examples mentioned here for the use-case.

Instead of designing the APIs with many overloaded optional parameters in the `GET` request, we should define a true REST API, which in your example would be:
```
/v1/deployment/<id>
-> Returns the deployment metadata excluding the containers.

/v1/deployment/<id>/containers
-> Returns the containers associated with the deployment.
```

I fear the optional parameters for the APIs might be too complex in the long-run as a pattern to follow. Personally feel the UX is also quite bad, comparing it to traditional REST API designs where each subresource to fetch is outlined in the path as a resource instead of a optional query parameter.

Reply:
Although your suggestions are valid endpoints, the biggest issue that the approach has is that users would be required to make too many API calls to compose what seems like one response object.